### PR TITLE
These were the files modified to implement the Abort feature.

### PR DESCRIPTION
**Pull request by iotsiabah**
Implemented the abort feature. The parameter
AbortAndResetOnTimeout Yes
enables abort on timeout and the parameter
ResetControllerOnTimeout Yes
enables reset controller on timeout.

### CODE REVIEW

File: include/spdk/nvme.h [hunk 1]
```diff
@@ -387,9 +387,10 @@ struct spdk_nvme_qpair;
  * detected on a request.
  */
 typedef void (*spdk_nvme_timeout_cb)(struct spdk_nvme_ctrlr *ctrlr,
-				     struct spdk_nvme_qpair *qpair,
-				     void *cb_arg);
-
+				     void *cb_arg,
+				     uint16_t abort_count,
+				     uint16_t cid,
+				     uint16_t sqid);
```
Comment (dverkamp-intel) [inline include/spdk/nvme.h]: I think passing back a spdk_nvme_qpair pointer rather than sqid would be best, and then we can make the abort command function take a qpair pointer as well.

File: include/spdk/nvme.h [hunk 2]
```diff
@@ -1141,6 +1142,36 @@ int spdk_nvme_ns_cmd_reservation_report(struct spdk_nvme_ns *ns,
...
+ *
+ * \param ctrlr controller to the Nvme device
+ * \param cid identification number of the command to be aborted.
+ * \param sqid submission queue number of the command to be aborted
+ * \param cb_fn a function pointer to the abort call back function
+ * \param cb_arg user argument to the abort command callback function
+ *
+ * \return 0 if successfully submitted, Less than 0 if it fails
+ *
+ * Note: The user must ensure to decrement the controller member
```
Comment (dverkamp-intel) [inline include/spdk/nvme.h]: I'm not sure this Note makes sense; why is the user responsible for this?  We should be able to handle this count internally in the NVMe library.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 3]
```diff
@@ -506,16 +507,62 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
...
+			 void *cb_arg,
+			 uint16_t timeout_count,
+			 uint16_t cid,
+			 uint16_t sqid)
 {
 	int rc;
 
-	SPDK_WARNLOG("Warning: Detected a timeout. ctrlr=%p qpair=%p\n", ctrlr, qpair);
+	if (cb_arg) {
+		SPDK_WARNLOG("Warning: Detected a timeout. user_data=%p ctrlr=%p \n", cb_arg, ctrlr);
```
Comment (dverkamp-intel) [inline lib/bdev/nvme/blockdev_nvme.c]: cb_arg is not used by the NVMe blockdev, so I don't think there's any reason to change this printout.

File: lib/nvme/nvme_ctrlr.c [hunk 4]
```diff
@@ -469,7 +469,7 @@ nvme_ctrlr_state_string(enum nvme_ctrlr_state state)
 {
 	switch (state) {
 	case NVME_CTRLR_STATE_INIT:
-		return "init";
+			return "init";
```
Comment (dverkamp-intel) [inline lib/nvme/nvme_ctrlr.c]: Unrelated whitespace change - please undo this.

File: include/spdk/nvme.h (hunk 2 above)
Comment (iotsiabah) [inline include/spdk/nvme.h]: Hi Daniel, there were two design alternatives if one chooses not to exceed the abort count limit:
1. the current implementation which places responsibility on the caller of 
    spdk_nvme_ctrlr_cmd_abort(..) to decrement  ctrlr->curr_abort_count 
    member in their abort callback function. 

2. The 2nd implementation would require the incrementation of  ctrlr->curr_abort_count to be
    localized as follows:
   a. acquire the lock before calling the abort callback function.
   b. If abort was performed in the abort callback function AND if the call to abort was successful,
       return a special code to indicate as such AND in order to increment the 
       ctrlr->curr_abort_count
       member. Otherwise, do not ctrlr->curr_abort_count member when abort callback function
       return.
   c. release the lock

   The 2nd implementation places no responsibility on the caller of spdk_nvme_ctrlr_cmd_abort(..)
   to decrement ctrlr->curr_abort_count member. However, there are two disadvantages, 1) the 
   callback function must return three different return types indicate that abort was called
   and the call succeeded, failed, or a controller reset was performed instead. 2) the lock will be
   acquired again in the abort function.

The 2nd choice makes the code reading difficult to understand so, we chose the current implementation.

**Comment by iotsiabah** [comment]
Hi Daniel, i made all the changes and all components do build successfully and i updated the pull request yesterday. However, on the pull request page under "All checks have failed", it is saying there is 1 failing check. The Details link on the right give no relevant  information. Anyway, the requested changes are in.  Thank you.

Isaac

**Comment by iotsiabah** [comment]
Hi Daniel, We found the problem and we are fixing it. Thanks.

**Comment by iotsiabah** [comment]
Hi Daniel, the update is completed and yesterday's problem is fixed. Thank you.

File: lib/nvme/nvme_pcie.c [hunk 5]
```diff
@@ -1841,7 +1841,14 @@ nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
 		 * Request has timed out. This could be i/o or admin request.
 		 * Call the registered timeout function for user to take action.
 		 */
-		ctrlr->timeout_cb_fn(ctrlr, qpair, ctrlr->timeout_cb_arg);
+
+		tr->req->timeout_count++;
+		tr->submit_tick = spdk_get_ticks() + ctrlr->timeout_ticks;
```
Comment (dverkamp-intel) [inline lib/nvme/nvme_pcie.c]: This should probably be resetting submit_tick to the current spdk_get_ticks() - the timeout_ticks addition should be moved to the place that is checking for timeout. (We modified the way timeouts are checked so that submission time can be recorded separately and used for I/O statistics.)

File: test/lib/nvme/unit/nvme_ctrlr_cmd_c/nvme_ctrlr_cmd_ut.c [hunk 6]
```diff
@@ -459,10 +459,18 @@ static void
 test_abort_cmd(void)
 {
 	struct spdk_nvme_ctrlr	ctrlr = {};
+	struct spdk_nvme_qpair	qpair = {};
+	ctrlr.curr_abort_count = 0;
 
 	verify_fn = verify_abort_cmd;
 
-	nvme_ctrlr_cmd_abort(&ctrlr, abort_cid, abort_sqid, NULL, NULL);
+	qpair.id = abort_cid;
```
Comment (dverkamp-intel) [inline test/lib/nvme/unit/nvme_ctrlr_cmd_c/nvme_ctrlr_cmd_ut.c]: The queue pair ID should not be assigned a command ID (those are different, unrelated things). I believe abort_sqid is what you want.

File: include/spdk/nvme.h (hunk 2 above)
Comment (dverkamp-intel) [inline include/spdk/nvme.h]: The timeout callback will always have the ctrlr_lock already held, since it is called in response to a spdk_nvme_process_admin_completions() function call, so the timeout completion handler code can assume it already has the lock.

It might be simplest to just not add the abort count limit checking in this first version; the worst that can happen is that the Abort command will be failed by the controller with Abort Command Limit Exceeded status, and then the abort completion handler can immediately issue a controller reset to recover.  (If enough commands time out simultaneously that the abort limit is exceeded, the controller is probably not responsive anyway.)

If we do want to enforce the abort limit, then I still think it should be entirely handled inside the NVMe library.  The simplest way to do this would be to have an intermediate callback function that does the decrement internally.  An example where the NVMe library already does this sort of intermediate callback is the nvme_user_copy_cmd_complete() path; you can reuse the existing user_cb_fn and user_cb_arg fields to store the actual user abort callback.

File: lib/nvme/nvme_ctrlr_cmd.c [hunk 7]
```diff
@@ -408,7 +421,10 @@ nvme_ctrlr_cmd_abort(struct spdk_nvme_ctrlr *ctrlr, uint16_t cid,
 	cmd->opc = SPDK_NVME_OPC_ABORT;
 	cmd->cdw10 = (cid << 16) | sqid;
 
-	return nvme_ctrlr_submit_admin_request(ctrlr, req);
+	rc = nvme_ctrlr_submit_admin_request(ctrlr, req);
+	ctrlr->curr_abort_count++;
```
Comment (dverkamp-intel) [inline lib/nvme/nvme_ctrlr_cmd.c]: curr_abort_count should probably only be incremented if the nvme_ctrlr_submit_admin_request() call was successful.

File: include/spdk/nvme.h [hunk 8]
```diff
@@ -1141,6 +1142,36 @@ int spdk_nvme_ns_cmd_reservation_report(struct spdk_nvme_ns *ns,
 					void *payload, uint32_t len,
 					spdk_nvme_cmd_cb cb_fn, void *cb_arg);
 
+/**
+ * \brief Sends an abort command to the Nvme device for the request indicated
+ *  by cid and sqid
+ *
+ * \param ctrlr controller to the Nvme device
+ * \param cid identification number of the command to be aborted.
+ * \param sqid submission queue number of the command to be aborted
```
Comment (dverkamp-intel) [inline include/spdk/nvme.h]: The sqid parameter was changed to a qpair pointer, so this documentation needs to be updated.

File: include/spdk/nvme.h [hunk 9]
```diff
@@ -1141,6 +1142,36 @@ int spdk_nvme_ns_cmd_reservation_report(struct spdk_nvme_ns *ns,
 					void *payload, uint32_t len,
 					spdk_nvme_cmd_cb cb_fn, void *cb_arg);
 
+/**
+ * \brief Sends an abort command to the Nvme device for the request indicated
```
Comment (dverkamp-intel) [inline include/spdk/nvme.h]: Nvme -> NVMe

**Comment by iotsiabah** [comment]
Hi Daniel, i have completed making all the suggested changes in your yesterday's email. Thank you.

Isaac

**Comment by iotsiabah** [comment]
Hi Daniel, how are you? i completed all the suggested changes.

**Comment by dverkamp-intel** [comment]
Hi Isaac, sorry for the lack of feedback - I've split up some of the prep work for abort into separate patches and merged a couple, but the blockdev_nvme change is still pending review.  We are trying to think through the implications of the timeout_count change - I'm not totally sure I understand what it is trying to accomplish.  It seems to me that the ideal behavior would be to issue an abort when a command times out, followed by a full controller reset if the abort command itself times out, rather than trying to count timeouts of individual commands.  A simple approximation of this could be issuing a controller reset if any timeout occurs for a command on an admin queue pair, since we have the qpair information in the timeout callback already.

**Comment by pvonstamwitz** [comment]
The full controller reset will follow the abort if the abort command times out or if the abort command finishes (most likely with “command was not aborted” status) and the aborted command times out again. It covers the case where a command somehow got “lost”.

**Comment by iotsiabah** [comment]
Hi Daniel, we agree with your suggestion of resetting the controller when a abort command timed out. I have made the changes and have updated the pull request. Thank you.

**Comment by iotsiabah** [comment]
Hi Daniel, i have updated the pull request to include a default timeout in case the user registers for timeout but do not specify a timeout value. Thank you.

File: etc/spdk/iscsi.conf.in [hunk 10]
```diff
@@ -105,8 +105,12 @@
...
   # claim all of them.
   NumControllers 2
-  # Registers the application to receive timeout callback and to reset the controller.
-  ResetControllerOnTimeout Yes
+  # Registers the application to receive a callback in the event of a command timeout.
+  # Users may specify whether the callback will attempt an abort before resetting the
+  # controller (if the abort fails) or to simply reset the controller. If more than one
+  # option is set, the least intrusive (i.e. abort) takes precedent.
+  AbortAndResetOnTimeout Yes
+  ResetControllerOnTimeout No
```
Comment (jimharris) [inline etc/spdk/iscsi.conf.in]: I think a single tri-state variable would be better.  Maybe something like:

HandleTimeouts None/Abort/Reset

Then we do not have to arbitrate (and explain) what happens when user selects Yes for both.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 11]
```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
 }
 
 static void
-blockdev_nvme_timeout_cb(void *cb_arg, struct spdk_nvme_ctrlr *ctrlr,
-			 struct spdk_nvme_qpair *qpair, uint16_t cid)
+abort_cb(void *arg, const struct spdk_nvme_cpl *cpl)
```
Comment (jimharris) [inline lib/bdev/nvme/blockdev_nvme.c]: I do not think we should add an empty abort_cb() routine.  Either we should actually do a real abort here, or we should not indicate that the driver will do aborts if the proper mode is set.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 12]
```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
...
+		/*
+		 * Abort on timeout AND it is the first time timeout has
+		 * expired for this request.
+		 */
+
+		rc = spdk_nvme_ctrlr_cmd_abort(ctrlr, qpair, cid, abort_cb, ctrlr);
+		if (rc) {
+			/**
+			 *  Above function can fail for one of these conditions:
+			 * 1. Memmory allocation error
```
Comment (jimharris) [inline lib/bdev/nvme/blockdev_nvme.c]: Memmory => Memory

File: lib/bdev/nvme/blockdev_nvme.c [hunk 13]
```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
...
+			 * 2. We reached the acl (abort count limit) limit  for abort
```
Comment (jimharris) [inline lib/bdev/nvme/blockdev_nvme.c]: extra space between "limit" and "for"

File: lib/nvme/nvme_ctrlr.c [hunk 14]
```diff
@@ -1465,6 +1465,7 @@ void
 spdk_nvme_ctrlr_register_timeout_callback(struct spdk_nvme_ctrlr *ctrlr,
 		uint32_t nvme_timeout, spdk_nvme_timeout_cb cb_fn, void *cb_arg)
 {
+	nvme_timeout = nvme_timeout == 0? NVME_IO_TIMEOUT:nvme_timeout;
```
Comment (jimharris) [inline lib/nvme/nvme_ctrlr.c]: Extra spaces needed before the question mark and before and after the colon.

File: lib/nvme/nvme_pcie.c [hunk 15]
```diff
@@ -1841,7 +1841,24 @@ nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
...
+		} else {
+			tr->req->timeout_count++;
+		}
+
+		tr->submit_tick = spdk_get_ticks();
+		ctrlr->timeout_cb_fn(ctrlr,
+				     qpair,
+				     ctrlr->timeout_cb_arg,
+				     tr->req->timeout_count,
+				     tr->req->cmd.cid);
```
Comment (jimharris) [inline lib/nvme/nvme_pcie.c]: Not really related to these patches, but it seems we should somehow make sure we do not call this timeout_cb_fn for internally generated requests?  For example, admin commands during initialization.  Something to consider later.

File: test/lib/nvme/unit/nvme_ctrlr_cmd_c/nvme_ctrlr_cmd_ut.c [hunk 16]
```diff
@@ -461,7 +461,7 @@ test_abort_cmd(void)
 	struct spdk_nvme_ctrlr	ctrlr = {};
 	struct spdk_nvme_qpair	qpair = {};
 
-	verify_fn = verify_abort_cmd;
+        verify_fn = verify_abort_cmd;
```
Comment (jimharris) [inline test/lib/nvme/unit/nvme_ctrlr_cmd_c/nvme_ctrlr_cmd_ut.c]: This should be reverted.

File: lib/nvme/nvme_ctrlr_cmd.c [hunk 17]
```diff
@@ -393,8 +393,11 @@ spdk_nvme_ctrlr_cmd_get_log_page(struct spdk_nvme_ctrlr *ctrlr, uint8_t log_page
 }
 
 int
-spdk_nvme_ctrlr_cmd_abort(struct spdk_nvme_ctrlr *ctrlr, struct spdk_nvme_qpair *qpair,
-			  uint16_t cid, spdk_nvme_cmd_cb cb_fn, void *cb_arg)
+spdk_nvme_ctrlr_cmd_abort(struct spdk_nvme_ctrlr *ctrlr, 
+			  struct spdk_nvme_qpair *qpair,
+			  uint16_t cid,
+			  spdk_nvme_cmd_cb cb_fn, 
+			  void *cb_arg)
```
Comment (jimharris) [inline lib/nvme/nvme_ctrlr_cmd.c]: Unrelated whitespace change - suggest making this part of a separate patch.

File: include/spdk/nvme.h [hunk 18]
```diff
@@ -1161,6 +1161,23 @@ int spdk_nvme_ns_cmd_reservation_report(struct spdk_nvme_ns *ns,
...
+ *  by cid
+ *
+ * \param ctrlr controller to the Nvme device
+ * \param qpair I/O queue pair associated with the request
+ * \param cid identification number of the command to be aborted.
+ * \param cb_fn a function pointer to the abort call back function
+ * \param cb_arg user argument to the abort command callback function
+ *
+ * \return 0 if successfully submitted, Less than 0 if it fails
+ */
```
Comment (benlwalker) [inline include/spdk/nvme.h]: This is trickier than it seems because abort is an ADMIN command and is sent on the ADMIN queue. Therefore, to actually see cb_fn called you have to poll the admin queue - not the qpair you are passing in. That's really confusing, so at a minimum it should be well documented here.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 19]
```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
...
+	/*
+	 * abort completion callback function. Currently, there is nothing to do
+	 * struct spdk_nvme_ctrlr *ctrlr = arg;
+	 */
+}
+
+static void
+blockdev_nvme_timeout_cb(struct spdk_nvme_ctrlr *ctrlr,
+			 struct spdk_nvme_qpair *qpair,
+			 void *cb_arg,
```
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: The timeout_count here is probably not the right way to present this. Instead, a command should only ever time out one time. On the first timeout, the user may elect to send an abort for that command. If that abort times out, this callback will be called for the abort command, and in response to a timeout of an abort command (or any ADMIN command really), a reset should be performed instead of sending another abort.

File: lib/bdev/nvme/blockdev_nvme.c [hunk 20]
```diff
@@ -502,16 +503,59 @@ probe_cb(void *cb_ctx, const struct spdk_nvme_transport_id *trid,
...
+abort_cb(void *arg, const struct spdk_nvme_cpl *cpl)
+{
+	/*
+	 * abort completion callback function. Currently, there is nothing to do
+	 * struct spdk_nvme_ctrlr *ctrlr = arg;
+	 */
+}
+
+static void
+blockdev_nvme_timeout_cb(struct spdk_nvme_ctrlr *ctrlr,
```
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: If an ADMIN command times out this qpair will point to the ADMIN queue. However, we intentionally do not expose the ADMIN queue through the public API. I think in the case of an ADMIN command timeout, qpair should just be NULL. This also provides a convenient way to check whether an I/O or ADMIN command timed out.

File: lib/nvme/nvme_internal.h [hunk 21]
```diff
@@ -95,6 +95,11 @@
 #define NVME_MAX_AER_LOG_SIZE		(4096)
 
 /*
+ * Admin and i/o commands default timeout in seconds
```
Comment (benlwalker) [inline lib/nvme/nvme_internal.h]: I/O

File: lib/nvme/nvme_pcie.c (hunk 15 above)
Comment (iotsiabah) [inline lib/nvme/nvme_pcie.c]: Jim, the code in nvme_pcie_qpair_process_completions(..) function checks to see if the controller state is NVME_CTRLR_STATE_READY before it calls the timeout_cb_fn function.

File: etc/spdk/iscsi.conf.in (hunk 10 above)
Comment (iotsiabah) [inline etc/spdk/iscsi.conf.in]: Agreed, it is a good idea and will implement it.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 11 above)
Comment (iotsiabah) [inline lib/bdev/nvme/blockdev_nvme.c]: OK, the empty abort_cb() function is removed, and i passed NULL as abort callback function instead.

File: include/spdk/nvme.h [hunk 22]
```diff
@@ -386,11 +386,11 @@ struct spdk_nvme_qpair;
  * Signature for the callback function invoked when a timeout is
  * detected on a request.
  */
-typedef void (*spdk_nvme_timeout_cb)(void *cb_arg,
-				     struct spdk_nvme_ctrlr *ctrlr,
+typedef void (*spdk_nvme_timeout_cb)(struct spdk_nvme_ctrlr *ctrlr,
 				     struct spdk_nvme_qpair *qpair,
+				     void *cb_arg,
```
Comment (iotsiabah) [inline include/spdk/nvme.h]: Ben, sorry, the abort_count name was a misnomer or typing error. The variable name is timeout_count and has been changed in the nvme.h.

File: include/spdk/nvme.h (hunk 18 above)
Comment (iotsiabah) [inline include/spdk/nvme.h]: The qpair is needed in spdk_nvme_ctrlr_cmd_abort(..) function so that the sqid can be retrieved. I had in the past passed the sqid into spdk_nvme_ctrlr_cmd_abort(..) but, was advised to pass the qpair instead.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 20 above)
Comment (iotsiabah) [inline lib/bdev/nvme/blockdev_nvme.c]: Ben, our Fujitsu Japan wants ADMIN commands (except ABORT command) that time out to be aborted, therefore, the qpair can not be NULL.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 19 above)
Comment (pvonstamwitz) [inline lib/bdev/nvme/blockdev_nvme.c]: Ben, you are correct. If the abort times out, the controller is reset. The timeout_count is used to handle the corner case where the abort finishes without cleaning up the original command, i.e. abort command returns with "command not aborted" or "abort command limit exceeded" status. For this case, we extend the timeout value for the original command and bump the timeout_count. If it times out again, it means that the abort did not work. In this case, we do not send another abort, but simply reset the controller.

**Comment by iotsiabah** [comment]
I will update this pull request with the above suggested changes tomorrow morning.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 19 above)
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: So the case the timeout_count is addressing is handling an abort failure. There are, of course, several ways the abort can fail and at least two of those errors may not indicate a serious problem (the two you listed). Below is a bit of analysis on what I think the "right" thing to do is in response to an abort failure.

The original I/O, which I'll refer to as O, timed out. Then, the software sent an abort, A. A then completes with an error.

If A completes with the status "command not aborted", there does not seem to be any reason to believe that sending an abort again will work correctly. O has also already timed out, so waiting more time is not valuable. Therefore, the only path forward seems to be to do a controller reset or delete the queue pair.

If A completes with the status "abort command limit exceeded", then there is reason to believe that waiting some period of time and then trying to abort again may work. I think we can handle this case internally in the NVMe driver by adding an abort command queue to the admin queue pair. The limit on simultaneous outstanding aborts is known at start up time, so we can manage the queue in software and effectively avoid this error.

If A completes with any other error, the controller should be reset or the queue pair should be deleted.

File: include/spdk/nvme.h (hunk 22 above)
Comment (pvonstamwitz) [inline include/spdk/nvme.h]: The timeout is extended by refreshing tr->submit_tick before calling the callback. If, for some reason, the abort does not clean up the command, this "second" timeout will indicate that something is wrong and that the controller should be reset.

File: include/spdk/nvme.h (hunk 18 above)
Comment (pvonstamwitz) [inline include/spdk/nvme.h]: The admin queue poller is registered in attach_cb. We can document that the admin queue poller needs to be registered if aborts are supported.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 20 above)
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: I'm actually going to sit down and do a few tests regarding aborting ADMIN commands. I'll update this when I have results.

Comment (pvonstamwitz) [inline lib/bdev/nvme/blockdev_nvme.c]: Internally, we have gone back and forth on this. Currently, the code will issue an abort for any command other than an abort. If an abort times out, then the controller is reset. We will extend this to any admin command for now.

I understand the use of aborts for speculative reads in a distributed system with multiple copies, but historically aborts have been used perform error recovery as a less than draconian way than a reset.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 19 above)
Comment (pvonstamwitz) [inline lib/bdev/nvme/blockdev_nvme.c]: As I commented previously, when a command times out, the timeout value is extended by refreshing tr->submit_ticks before calling the callback. This extension has two purposes. It prevents calling the callback over and over again on subsequent polls and it provides a fallback in case of abort failure.

At the callback function, timeout_count will either be 1 indicating that this is the initial timeout, or 2 indicating that the extended time has also elapsed.

At the callback with aborts enable:
If timeout_count = 1, the callback will issue an abort. 
If timeout_count > 1, this indicates that an abort was sent and the original I/O still did not complete within the extended time period. Therefore, the callback will reset the controller.

We actually do not do any error checking for the abort command other than checking for timeout (which results in an automatic reset.) All other errors will be covered by the timeout extension. (We actually had earlier code that avoided the "abort command limit exceeded" scenario - we can add this at a later PR.)

Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: I understand what you're doing with the submit_ticks and it will definitely work, my questions are all coming from the perspective of trying to find the best way to handle this for everyone using the NVMe driver. I'm thinking that instead of a timeout_count in the tracker, we should put a flag in the tracker for whether a tracker is considered "timed out". The timeout callback will only be called once for a tracker's lifetime then. I'd then move the code that previously ran when timeout_count was greater than 1 to the abort completion handler. If the abort fails, reset the controller.

Let's definitely add the handling for the abort command limit too, as a separate PR.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 20 above)
Comment (benlwalker) [inline lib/bdev/nvme/blockdev_nvme.c]: I just did a bunch of tests on an Intel P3700 and learned a little bit. For the ADMIN commands I tried (log pages, set features, etc.) they appear to execute serially. The device does not pick up the next ADMIN command from the queue until the previous one has completed. That means aborting an ADMIN command is impossible (because abort is an ADMIN command and won't be seen until the previous commands complete). This is of course implementation specific, and I didn't send every possible ADMIN command, but it's evidence that aborting ADMIN commands is not going to be a reliable strategy for error recovery across all NVMe devices.

Aborts on I/O commands, on the other hand, work fine. Therefore, I think the right strategy is to send an ABORT when an I/O command times out. If any ADMIN command times out, reset the controller.

File: lib/nvme/nvme_pcie.c [hunk 23]
```diff
@@ -1841,7 +1841,24 @@ nvme_pcie_qpair_check_timeout(struct spdk_nvme_qpair *qpair)
 		 * Request has timed out. This could be i/o or admin request.
 		 * Call the registered timeout function for user to take action.
 		 */
-		ctrlr->timeout_cb_fn(ctrlr->timeout_cb_arg, ctrlr, qpair, tr->cid);
+
```
Comment (benlwalker) [inline lib/nvme/nvme_pcie.c]: The logic in this function was fine when the user's only recourse for any time out was to do a full controller reset, but if we're going to allow users to send ABORTs for specific commands, this function needs to loop over the entire list of trackers and call the callback for each one that is timed out. Right now, it only calls the callback for the oldest one.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 19 above)
Comment (pvonstamwitz) [inline lib/bdev/nvme/blockdev_nvme.c]: I see what you mean, but I'd rather not remove the extension of the timeout. The extended timeout covers all possible error cases and acts as watchdog timer in a sense. Relying on the abort completion handler can be tricky, e.g. "command not aborted" may or may not be an error depending on the timing.

We can move the reset due to the extended timeout to the NVMe driver which would allow us to use the "timed out" flag and call the callback only once.

File: lib/nvme/nvme_pcie.c (hunk 23 above)
Comment (pvonstamwitz) [inline lib/nvme/nvme_pcie.c]: This implementation is designed for a single I/O timeout value on a channel basis. It is for error recovery and therefore designed for minimal performance impact. We are simply allowing the user to try an abort instead of a reset.

It would be a great addition to allow a user to send an abort for any specific command for any reason. This would require a one-time traversal of the tracker queue to match the user's context. But, this is orthogonal to what we are trying to do here.

Another idea would be to allow the user to specify timeout values on a per command basis. I'm not sure how useful this is, but I do know our testing showed that traversing the tracker queue, even periodically, has measurable impact on latency.

The great thing about the callback is that the user can turn it off. With the ability to abort specific commands (which we should do), the user is free to be as fine-grained as they want.

File: lib/bdev/nvme/blockdev_nvme.c (hunk 19 above)
Comment (iotsiabah) [inline lib/bdev/nvme/blockdev_nvme.c]: Ben, resetting the controller if the abort fails may not be desired here because when i tested the abort command, i found most of the time when abort is a issued,  "the command to abort may have already completed" and you get from the admin completion queue DWORD 0 bit 0 is set to 1 for the abort command. The status code and status code type are set to 0 because the abort command executed successfully although mission was not accomplished. 
This is abort command behavior but controller itself is fine, therefore, resetting the controller may not be desired.

**Comment by iotsiabah** [comment]
Hi Ben, i have made the changes according to the discussions we have had and have updated the Abort pull request. Thank you very much.

Isaac

**Comment by benlwalker** [comment]
Isaac - can you take a look at PR #134? It's very much based on your patches here with only a couple of differences around handling abort failures and abort time outs.

**Comment by iotsiabah** [comment]
Please, go ahead and merge them into the SPDK. Thank you.

**Comment by benlwalker** [comment]
Merged a related pull request that encompasses this one. Thank you so much for your submission!
